controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,895 @@
+const (
+	controllerName           = "aks-controller"
+	controllerRemoveName     = "aks-controller-remove"
+	aksConfigNotCreatedPhase = ""
+	aksConfigActivePhase     = "active"
+	aksConfigUpdatingPhase   = "updating"
+	aksConfigImportingPhase  = "importing"
+	aksClusterConfigKind     = "AKSClusterConfig"
+)
+
+type Handler struct {
+	aksCC           v10.AKSClusterConfigClient
+	aksEnqueueAfter func(namespace, name string, duration time.Duration)
+	aksEnqueue      func(namespace, name string)
+	secrets         wranglerv1.SecretClient
+	secretsCache    wranglerv1.SecretCache
+}
+
+func Register(
+	ctx context.Context,
+	secrets wranglerv1.SecretController,
+	aks v10.AKSClusterConfigController) {
+
+	controller := &Handler{
+		aksCC:           aks,
+		aksEnqueue:      aks.Enqueue,
+		aksEnqueueAfter: aks.EnqueueAfter,
+		secretsCache:    secrets.Cache(),
+		secrets:         secrets,
+	}
+
+	// Register handlers
+	aks.OnChange(ctx, controllerName, controller.recordError(controller.OnAksConfigChanged))
+	aks.OnRemove(ctx, controllerRemoveName, controller.OnAksConfigRemoved)
+}
+
+func (h *Handler) OnAksConfigChanged(key string, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfig, error) {
+	if config == nil || config.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	ctx := context.Background()
+	switch config.Status.Phase {
+	case aksConfigImportingPhase:
+		return h.importCluster(ctx, config)
+	case aksConfigNotCreatedPhase:
+		return h.createCluster(ctx, config)
+	case aksConfigActivePhase, aksConfigUpdatingPhase:
+		return h.checkCluster(ctx, config)
+	default:
+		return config, fmt.Errorf("invalid phase: %v", config.Status.Phase)
+	}
+}
+
+func (h *Handler) OnAksConfigRemoved(key string, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfig, error) {
+	if config.Spec.Imported {
+		logrus.Infof("Cluster [%s] is imported, will not delete AKS cluster", config.Spec.DisplayName)
+		return config, nil
+	}
+
+	ctx := context.Background()
+	logrus.Infof("Removing cluster [%s] for config [%s]", config.Spec.ClusterName, config.Spec.DisplayName)
+
+	azureAuthorizer, err := newClientAuthorizer(h.secretsCache, config.Spec)
+	if err != nil {
+		return config, err
+	}
+	resourceGroupsClient, err := newResourceGroupsClient(azureAuthorizer, config.Spec)
+	if err != nil {
+		return config, err
+	}
+	resourceClusterClient, err := newClustersClient(azureAuthorizer, config.Spec)
+	if err != nil {
+		return config, err
+	}
+
+	if existsCluster(ctx, resourceClusterClient, config.Spec) {
+		if err := removeCluster(ctx, resourceClusterClient, config.Spec); err != nil {
+			return config, fmt.Errorf("error removing cluster [%s] message %v", config.Spec.ClusterName, err)
+		}
+	}
+
+	logrus.Infof("Removing resource group [%s] for config [%s]", config.Spec.ResourceGroup, config.Spec.DisplayName)
+
+	if existsResourceGroup(ctx, resourceGroupsClient, config.Spec.ResourceGroup) {
+		if err := removeResourceGroup(ctx, resourceGroupsClient, config.Spec); err != nil {
```

Comment:
Why would we delete the resource group? You can delete a cluster without deleting the resource group the cluster is in, no?

Reply:
Because `resource group` is needed to create AKS and we create it every time. So it has to be cleaned up afterwards